### PR TITLE
Remove stale SAM3 decoder coordinate caches

### PR DESCRIPTION
**Pull request by tritsystem**
Changing a warmed SAM3 decoder to another dtype or device could reuse stale coordinate tensors and fail in the RPB projections. Build the two coordinate vectors from the current reference boxes and pass the existing integer feature sizes from the image model, removing both caches, their invalidation state, and the encoder shape tensors that no longer have consumers. Compilation now starts at construction, as in the encoder, because it no longer needs a cache-warming forward.

Validation: exact FP32 parity with the previous decoder at square and rectangular feature sizes; FP16/BF16/FP32 transitions; CPU→MPS→CPU; Inductor fullgraph RPB and fullgraph encoder/decoder checks; training gradients; and loading a pre-change serialized decoder with a warmed cache before converting to FP16. CPU RPB medians at 200 queries, 256 channels, and a 72×72 feature map were 6.096 ms before and 6.102 ms after. Real pretrained `sam3.pt` inference through `SAM3Predictor` in a downstream annotation service also produced identical baseline/candidate polygons and boxes for positive/negative prompts, repeated clicks, and CPU feature-cache restoration after eviction. CUDA was not run locally.

The final change removes 60 net lines and the constructor-bypassing test fixture. Focused validation uses the real encoder and decoder constructors; no new test code is added.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tritsystem, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/sam/sam3/decoder.py [hunk 1]
```diff
@@ -317,22 +317,27 @@ def _get_rpb_matrix(self, reference_boxes, feat_size):
...
         if self.compilable_cord_cache is None:
             self.compilable_cord_cache = self._get_coords(H, W, reference_boxes.device, reference_boxes.dtype)
-            self.compilable_stored_size = (H, W)
+            self.compilable_cache_key = cache_key
 
-        if torch.compiler.is_dynamo_compiling() or self.compilable_stored_size == (
-            H,
-            W,
-        ):
+        if torch.compiler.is_dynamo_compiling() or self.compilable_cache_key == cache_key:
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/sam3/decoder.py]: 💡 **MEDIUM**: Loading a pre-change Ultralytics checkpoint deserializes the saved EMA module directly, without rerunning `__init__`. An older `TransformerDecoder` therefore has `compilable_stored_size` but no `compilable_cache_key`, so its first eager `_get_rpb_matrix` call raises `AttributeError` instead of rebuilding the new cache. Treat a missing key as a cache miss (or migrate the legacy attribute) to preserve compatibility with existing SAM3 checkpoints.

**Suggested change:**
```suggestion
        if torch.compiler.is_dynamo_compiling() or getattr(self, "compilable_cache_key", None) == cache_key:
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed both changed files and the surrounding cache, forward, checkpoint, and test paths. The dtype/device cache fix is otherwise correctly scoped, but renaming a serialized module attribute breaks older full-module checkpoints unless the missing attribute is handled as a cache miss.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/models/sam/sam3/decoder.py:332` Loading a pre-change Ultralytics checkpoint deserializes the saved EMA module directly, without rerunning `__init__`. An older `TransformerDecoder` therefore has `compilable_stored_size` but no `compilable_cache_key`, so its first eager `_get_rpb_matrix` call raises `AttributeError` instead of rebuilding the new cache. Treat a missing key as a cache miss (or migrate the legacy attribute) to preserve compatibility with existing SAM3 checkpoints.
</details>

**Comment by tritsystem** [comment]
I have read the CLA Document and I sign the CLA

**Review by glenn-jocher — APPROVED** [review]
LGTM at 293b0b82d5344cb0734b4556e55265b7bdc59116. Completed a cold full-diff review against AGENTS.md; independent Claude Code `claude-fable-5-1` review also returns LGTM at this exact head.

The final diff removes 60 net lines: both stale coordinate caches, the warmup compilation state, and unused encoder shape outputs. Real encoder/decoder parity, compiled execution, dtype/device transitions, training gradients, and a pre-change warmed-module checkpoint all passed. No new test fixture or compatibility path.

CI is still running; this approval is a code review, not a merge authorization.

**Comment by glenn-jocher** [comment]
Thank you for identifying and fixing the SAM3 dtype/device cache bug. Merged after simplifying the repair to remove the caches and unused shape plumbing (60 net lines removed), independent Claude Fable 5.1 high reviews, real encoder/decoder and compilation checks, and real pretrained SAM3 annotation-service inference. Positive/negative prompts, repeated clicks, and feature-cache restoration produced exactly the same outputs as the baseline. All CI checks completed green.
